lib/Serialization/Deserialization.cpp

```diff
@@ -4132,7 +4124,20 @@ llvm::Error DeclDeserializer::deserializeDeclAttributes() {
     unsigned recordID = MF.fatalIfUnexpected(
         MF.DeclTypeCursor.readRecord(entry.ID, scratch, &blobData));
 
-    if (isDeclAttrRecord(recordID)) {
+    if (recordID == ERROR_FLAG) {
+      IsInvalid = true;
+      if (ctx.LangOpts.AllowModuleWithCompilerErrors) {
+        if (!ctx.hadError()) {
+          ctx.Diags.diagnose(
```

Comment:
You could move the diagnostic emission at the point where the bit is set (`decl->setInvalidBit()`) in which case the decl name should be available for including in the diagnostic.
It would be useful to see in a build log what declaration is the invalid one.

Reply:
Good point, should have done that when I added `IsInvalid` - thanks.